# Exercici 1

### Descarrega el data set Airlines Delay: Airline on-time statistics and delay causes i carrega’l a un Pandas Dataframe. Explora les dades que conté, explica breument quines variables hi ha i queda’t únicament amb les columnes que consideris rellevants. Justifica la teva elecció.

In [68]:
import pandas as pd
import numpy as np

df_vuelos=pd.read_csv('DelayedFlights.csv')

df_vuelos.head()


Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0           0  2008      1           3          4   2003.0        1955   
1           1  2008      1           3          4    754.0         735   
2           2  2008      1           3          4    628.0         620   
3           4  2008      1           3          4   1829.0        1755   
4           5  2008      1           3          4   1940.0        1915   

   ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
0   2211.0        2225            WN  ...     4.0     8.0          0   
1   1002.0        1000            WN  ...     5.0    10.0          0   
2    804.0         750            WN  ...     3.0    17.0          0   
3   1959.0        1925            WN  ...     3.0    10.0          0   
4   2121.0        2110            WN  ...     4.0    10.0          0   

   CancellationCode  Diverted  CarrierDelay  WeatherDelay NASDelay  \
0                 N         0           NaN           NaN      NaN   
1                 N         0           NaN           NaN      NaN   
2                 N         0           NaN           NaN      NaN   
3                 N         0           2.0           0.0      0.0   
4                 N         0           NaN           NaN      NaN   

  SecurityDelay  LateAircraftDelay  
0           NaN                NaN  
1           NaN                NaN  
2           NaN                NaN  
3           0.0               32.0  
4           NaN                NaN  

[5 rows x 30 columns]

In [69]:
df_vuelos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1936758 entries, 0 to 1936757
Data columns (total 30 columns):
 #   Column             Dtype  
---  ------             -----  
 0   Unnamed: 0         int64  
 1   Year               int64  
 2   Month              int64  
 3   DayofMonth         int64  
 4   DayOfWeek          int64  
 5   DepTime            float64
 6   CRSDepTime         int64  
 7   ArrTime            float64
 8   CRSArrTime         int64  
 9   UniqueCarrier      object 
 10  FlightNum          int64  
 11  TailNum            object 
 12  ActualElapsedTime  float64
 13  CRSElapsedTime     float64
 14  AirTime            float64
 15  ArrDelay           float64
 16  DepDelay           float64
 17  Origin             object 
 18  Dest               object 
 19  Distance           int64  
 20  TaxiIn             float64
 21  TaxiOut            float64
 22  Cancelled          int64  
 23  CancellationCode   object 
 24  Diverted           int64  
 25  CarrierDelay      

In [70]:
# reducimos los registros a analizar a 200.000
df_vuelos_red = df_vuelos.sample(n=200000)
df_vuelos_red.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 200000 entries, 1230055 to 507388
Data columns (total 30 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   Unnamed: 0         200000 non-null  int64  
 1   Year               200000 non-null  int64  
 2   Month              200000 non-null  int64  
 3   DayofMonth         200000 non-null  int64  
 4   DayOfWeek          200000 non-null  int64  
 5   DepTime            200000 non-null  float64
 6   CRSDepTime         200000 non-null  int64  
 7   ArrTime            199296 non-null  float64
 8   CRSArrTime         200000 non-null  int64  
 9   UniqueCarrier      200000 non-null  object 
 10  FlightNum          200000 non-null  int64  
 11  TailNum            200000 non-null  object 
 12  ActualElapsedTime  199175 non-null  float64
 13  CRSElapsedTime     199989 non-null  float64
 14  AirTime            199175 non-null  float64
 15  ArrDelay           199175 non-null  float64
 

### Breve descripción de las columnas

- Unnamed: Indice de los registros.
- Year: Año del registro.	
- Month: Mes del registro en formato numérico.
- DayofMonth: Día del mes del registro en formato numérico.
- DayOfWeek: Día de la semana del registro en formato numérico.
- DepTime: Hora local de salida del vuelo (hhmm).
- CRSDepTime: Hora local prevista de salida del vuelo (hhmm).
- ArrTime: Hora local de llegada del vuelo (hhmm).
- CRSArrTime: Hora local prevista de llegada del vuelo(hhmm).
- UniqueCarrier: Identificación abreviada de la compañia.
- FlightNum: Número de vuelo.
- TailNum: Número de cola del vuelo.
- ActualElapsedTime: Tiempo transcurrido de vuelo en minutos.
- CRSElapsedTime: Tiempo previsto de vuelo en minutos.
- AirTime: Tiempo de vuelo en minutos( en aire).
- ArrDelay: Tiempo de retraso llegada en minutos.
- DepDelay: Tiempo de retraso salida en minutos.
- Origin: Código aeropuerto de origen.
- Dest: Código aeropuerto de destino.
- Distance: Distancia en millas.
- TaxiIn: Rodaje del avión puerta embarque-pista de despegue en minutos.
- TaxiOut: Rodaje del avión pista de despegue-puerta de embarque en minutos.
- Cancelled: Indicador de vuelo cancelado (1= yes, 0= no).
- CancellationCode: Razón de la cancelación (A = carrier, B = weather, C = NAS, D = security).
- Diverted: Indicador de vuelo desviado (1= yes, 0= no).
- CarrierDelay: Retraso causa Carrier en minutos.
- WeatherDelay: Retraso causa Weather en minutos.
- NASDelay: Retraso causa NAS en minutos.
- SecurityDelay: Retraso causa Security en minutos.
- LateAircraftDelay: Retraso causa LateAircraft en minutos.

### Columnas a eliminar y explicación

- Unnamed: 0, es una columna con información repetida.
- Year, Month, DayofMonth las unificamos en una nueva columna llamada 'Date'
- DepTime, CRSDepTime, ArrTime y CRSArrTime se pueden borrar ya que ya tenemos columnas que nos indican el tiempo empleado en los vuelos (ActualElapsedTime y CRSElapsedTime).
- Podemos eliminar también los campos TaxiIn y TaxiOut ya que no son relevantes en el cálculo total del retraso.
- CancellationCode se puede borrar, ya que para los motivos del ejercicio nos basta saber si el vuelo ha sido cancelado, pero no necesitamos la causa.

In [71]:
# unificamos Year, Month, DayofMonth en Date y colocamos esta última al principio de las columnas

df_vuelos_red['Date']=pd.to_datetime(dict(year=df_vuelos['Year'], month=df_vuelos['Month'], day=df_vuelos['DayofMonth']))
df_vuelos_red.insert(1, 'Date', df_vuelos_red.pop('Date'))

In [72]:
df_vuelos_red.head()

Unnamed: 0       Date  Year  Month  DayofMonth  DayOfWeek  DepTime  \
1230055     4122415 2008-07-16  2008      7          16          3   1518.0   
1760682     6531537 2008-12-23  2008     12          23          2   2302.0   
1302979     4363445 2008-08-15  2008      8          15          5   1716.0   
1235347     4138259 2008-07-31  2008      7          31          4   1608.0   
585395      1822696 2008-04-12  2008      4          12          6   1736.0   

         CRSDepTime  ArrTime  CRSArrTime  ... TaxiIn  TaxiOut Cancelled  \
1230055        1425   1921.0        1840  ...    6.0     16.0         0   
1760682        2015      4.0        2130  ...    6.0      7.0         0   
1302979        1700   2201.0        2153  ...    5.0     12.0         0   
1235347        1555   1742.0        1725  ...    7.0     17.0         0   
585395         1725   1909.0        1910  ...    4.0      8.0         0   

         CancellationCode  Diverted  CarrierDelay  WeatherDelay  NASDelay  \
1230055                 N         0          10.0           0.0       0.0   
1760682                 N         0         154.0           0.0       0.0   
1302979                 N         0           NaN           NaN       NaN   
1235347                 N         0          13.0           0.0       4.0   
585395                  N         0           NaN           NaN       NaN   

        SecurityDelay LateAircraftDelay  
1230055           0.0              31.0  
1760682           0.0               0.0  
1302979           NaN               NaN  
1235347           0.0               0.0  
585395            NaN               NaN  

[5 rows x 31 columns]

In [73]:
# borramos las columnas que no necesitamos
df_vuelos_red.drop(['Unnamed: 0','Year','Month','DayOfWeek','DayofMonth','DepTime','CRSDepTime','ArrTime','CRSArrTime','CancellationCode','TaxiIn','TaxiOut'],axis=1,inplace=True)

In [74]:
df_vuelos_red.head()

Date UniqueCarrier  FlightNum TailNum  ActualElapsedTime  \
1230055 2008-07-16            AA       1240  N632AA              183.0   
1760682 2008-12-23            WN       2610  N735SA               62.0   
1302979 2008-08-15            XE         22  N11199              165.0   
1235347 2008-07-31            AA       1827  N442AA              214.0   
585395  2008-04-12            WN       3212  N787SA               93.0   

         CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  \
1230055           195.0    161.0      41.0      53.0    ORD  MIA      1197   
1760682            75.0     49.0     154.0     167.0    SJC  LAX       308   
1302979           173.0    148.0       8.0      16.0    ONT  TUL      1236   
1235347           210.0    190.0      17.0      13.0    DFW  SMF      1431   
585395            105.0     81.0      -1.0      11.0    TPA  RDU       587   

         Cancelled  Diverted  CarrierDelay  WeatherDelay  NASDelay  \
1230055          0         0          10.0           0.0       0.0   
1760682          0         0         154.0           0.0       0.0   
1302979          0         0           NaN           NaN       NaN   
1235347          0         0          13.0           0.0       4.0   
585395           0         0           NaN           NaN       NaN   

         SecurityDelay  LateAircraftDelay  
1230055            0.0               31.0  
1760682            0.0                0.0  
1302979            NaN                NaN  
1235347            0.0                0.0  
585395             NaN                NaN

# Exercici 2

## Fes un informe complet del dataset:

### - Resumeix estadísticament el dataset i les columnes d’interès. Fes una anàlisi estadístic del que consideris rellevant.



In [75]:
round(df_vuelos_red.describe(),2)

FlightNum  ActualElapsedTime  CRSElapsedTime    AirTime   ArrDelay  \
count  200000.00          199175.00       199989.00  199175.00  199175.00   
mean     2185.97             133.41          134.41     108.37      42.21   
std      1941.13              72.36           71.57      68.90      56.53   
min         1.00              18.00            1.00       0.00     -70.00   
25%       616.00              80.00           82.00      58.00       9.00   
50%      1545.00             116.00          116.00      90.00      24.00   
75%      3429.00             165.00          165.00     137.00      56.00   
max      7829.00             724.00          660.00     654.00    1395.00   

        DepDelay   Distance  Cancelled   Diverted  CarrierDelay  WeatherDelay  \
count  200000.00  200000.00  200000.00  200000.00     128934.00     128934.00   
mean       43.19     766.67       0.00       0.00         19.18          3.59   
std        53.05     576.58       0.02       0.06         43.01         20.46   
min         6.00      31.00       0.00       0.00          0.00          0.00   
25%        12.00     341.00       0.00       0.00          0.00          0.00   
50%        24.00     606.00       0.00       0.00          2.00          0.00   
75%        53.00     998.00       0.00       0.00         21.00          0.00   
max      1403.00    4962.00       1.00       1.00       1348.00       1297.00   

        NASDelay  SecurityDelay  LateAircraftDelay  
count  128934.00      128934.00          128934.00  
mean       15.04           0.08              25.37  
std        34.15           1.79              42.44  
min         0.00           0.00               0.00  
25%         0.00           0.00               0.00  
50%         2.00           0.00               8.00  
75%        15.00           0.00              33.00  
max       919.00         167.00            1000.00

### Del análisis estadísticos de los datos sobre la muestra aleatoria de 200.000 registros, podemos resaltar lo siguiente:

- Que el vuelo que cubre la distancia máxima recorre 4962 millas y la distancia mínima es de 31 millas. 
- El vuelo más largo en tiempo ha durado 724 minutos y el más corto 18 minutos. 
- El mayor retraso en llegada ha sido de 1395 minutos y en salida de 1403 minutos. 
- Teniendo en cuenta que el retraso medio es de 42 y 43 minutos para llegada y salida respectivamente, podemos afirmar que tenemos claramente valores outliers en estas columnas.

### - Troba quantes dades faltants hi ha per columna.

In [76]:
df_vuelos_red.isnull().sum()

Date                     0
UniqueCarrier            0
FlightNum                0
TailNum                  0
ActualElapsedTime      825
CRSElapsedTime          11
AirTime                825
ArrDelay               825
DepDelay                 0
Origin                   0
Dest                     0
Distance                 0
Cancelled                0
Diverted                 0
CarrierDelay         71066
WeatherDelay         71066
NASDelay             71066
SecurityDelay        71066
LateAircraftDelay    71066
dtype: int64

Como el objetivo del ejercicio es calcular una serie de eventos es los que es necesario que el vuelo se haya producido, calculamos cuantos de los registros corresponden a vuelos cancelados o desviados y los eliminamos del datset.

In [77]:
df_vuelos_red.apply(lambda x: x['Cancelled'] > 0, axis=1).sum()

52

In [78]:
df_vuelos_red.apply(lambda x: x['Diverted'] > 0, axis=1).sum()

773

Podemos comprobar que la suma de ambos valores corresponde con los campos sin datos en las columnas ActualElapsedTime, AirTime y ArrDelay

In [79]:
df_vuelos_red = df_vuelos_red.drop(df_vuelos_red[df_vuelos_red['Cancelled']>0].index)
df_vuelos_red = df_vuelos_red.drop(df_vuelos_red[df_vuelos_red['Diverted']>0].index)

df_vuelos_red.apply(lambda x: x['Cancelled'] > 0, axis=1).sum()

0

In [80]:
df_vuelos_red.apply(lambda x: x['Diverted'] > 0, axis=1).sum()

0

In [81]:
df_vuelos_red.isnull().sum()

Date                     0
UniqueCarrier            0
FlightNum                0
TailNum                  0
ActualElapsedTime        0
CRSElapsedTime           0
AirTime                  0
ArrDelay                 0
DepDelay                 0
Origin                   0
Dest                     0
Distance                 0
Cancelled                0
Diverted                 0
CarrierDelay         70241
WeatherDelay         70241
NASDelay             70241
SecurityDelay        70241
LateAircraftDelay    70241
dtype: int64

Para las columnas CarrierDelay, WeatherDelay, NASDelay , SecurityDelay, LateAircraftDelay que no tienen datos, relleno los mismos con el valor 0 para que no interfieran en los futuros cálculos en el caso de ser necesarias. De esta manera el dataset queda limpio.

In [82]:
df_vuelos_red=df_vuelos_red.fillna(0)
df_vuelos_red.isnull().sum()

Date                 0
UniqueCarrier        0
FlightNum            0
TailNum              0
ActualElapsedTime    0
CRSElapsedTime       0
AirTime              0
ArrDelay             0
DepDelay             0
Origin               0
Dest                 0
Distance             0
Cancelled            0
Diverted             0
CarrierDelay         0
WeatherDelay         0
NASDelay             0
SecurityDelay        0
LateAircraftDelay    0
dtype: int64

### Crea columnes noves (velocitat mitjana del vol, si ha arribat tard o no...).

In [83]:
# Calculamos la velocidad media del vuelo en mph cómo la división entre la distancia ('Distance') y el tiempo del avión 
# en el aire ('Airtime'). Colocamos la nueva columna junto a la Distancia.


df_vuelos_red['Ave_Speed'] = round(df_vuelos_red['Distance']/ (df_vuelos_red['AirTime'] / 60),2)
df_vuelos_red.insert(12, 'Ave_Speed', df_vuelos_red.pop('Ave_Speed'))
df_vuelos_red.head()

Date UniqueCarrier  FlightNum TailNum  ActualElapsedTime  \
1230055 2008-07-16            AA       1240  N632AA              183.0   
1760682 2008-12-23            WN       2610  N735SA               62.0   
1302979 2008-08-15            XE         22  N11199              165.0   
1235347 2008-07-31            AA       1827  N442AA              214.0   
585395  2008-04-12            WN       3212  N787SA               93.0   

         CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  \
1230055           195.0    161.0      41.0      53.0    ORD  MIA      1197   
1760682            75.0     49.0     154.0     167.0    SJC  LAX       308   
1302979           173.0    148.0       8.0      16.0    ONT  TUL      1236   
1235347           210.0    190.0      17.0      13.0    DFW  SMF      1431   
585395            105.0     81.0      -1.0      11.0    TPA  RDU       587   

         Ave_Speed  Cancelled  Diverted  CarrierDelay  WeatherDelay  NASDelay  \
1230055     446.09          0         0          10.0           0.0       0.0   
1760682     377.14          0         0         154.0           0.0       0.0   
1302979     501.08          0         0           0.0           0.0       0.0   
1235347     451.89          0         0          13.0           0.0       4.0   
585395      434.81          0         0           0.0           0.0       0.0   

         SecurityDelay  LateAircraftDelay  
1230055            0.0               31.0  
1760682            0.0                0.0  
1302979            0.0                0.0  
1235347            0.0                0.0  
585395             0.0                0.0

In [84]:
# Comprobamos si el avión ha llegado tarde o no en función del valor del campo 'ArrDelay'. 
# Si este es >= 0 consideramos que sí ha llegado tarde. Colocamos la nueva columna junto a la columna "ArrDelay".

df_vuelos_red['Delayed'] = np.where(df_vuelos_red['ArrDelay'] >= 0, 1, 0)
df_vuelos_red.insert(8, 'Delayed', df_vuelos_red.pop('Delayed'))
df_vuelos_red.head()

Date UniqueCarrier  FlightNum TailNum  ActualElapsedTime  \
1230055 2008-07-16            AA       1240  N632AA              183.0   
1760682 2008-12-23            WN       2610  N735SA               62.0   
1302979 2008-08-15            XE         22  N11199              165.0   
1235347 2008-07-31            AA       1827  N442AA              214.0   
585395  2008-04-12            WN       3212  N787SA               93.0   

         CRSElapsedTime  AirTime  ArrDelay  Delayed  DepDelay  ... Dest  \
1230055           195.0    161.0      41.0        1      53.0  ...  MIA   
1760682            75.0     49.0     154.0        1     167.0  ...  LAX   
1302979           173.0    148.0       8.0        1      16.0  ...  TUL   
1235347           210.0    190.0      17.0        1      13.0  ...  SMF   
585395            105.0     81.0      -1.0        0      11.0  ...  RDU   

        Distance  Ave_Speed  Cancelled  Diverted  CarrierDelay  WeatherDelay  \
1230055     1197     446.09          0         0          10.0           0.0   
1760682      308     377.14          0         0         154.0           0.0   
1302979     1236     501.08          0         0           0.0           0.0   
1235347     1431     451.89          0         0          13.0           0.0   
585395       587     434.81          0         0           0.0           0.0   

         NASDelay  SecurityDelay  LateAircraftDelay  
1230055       0.0            0.0               31.0  
1760682       0.0            0.0                0.0  
1302979       0.0            0.0                0.0  
1235347       4.0            0.0                0.0  
585395        0.0            0.0                0.0  

[5 rows x 21 columns]

### Fes una taula de les aerolínies amb més endarreriments acumulats.

In [85]:
# Creamos una tabla temporal donde añadimos el campo compañia y el valor de 'ArrDelay'. En esta tabla quitamos aquellos valores
# donde retrasado = 0, ya que 'ArrDelay' tienen valores negativo o cero y los agrupamos por compañia sumando el valor de 'ArrayDelay'.
# Finalmente renombramos las columnas 'Delayed' y 'ArrDelay'
df_vuelos_red_comp = df_vuelos_red.loc[:,['UniqueCarrier','ArrDelay','Delayed']]
df_vuelos_red_comp = df_vuelos_red_comp.drop(df_vuelos_red_comp[df_vuelos_red_comp['Delayed']==0].index).groupby(['UniqueCarrier']).sum().sort_values('ArrDelay', ascending=False)
df_vuelos_red_comp.rename(columns = {'ArrDelay':'AcumDelay'}, inplace = True)
df_vuelos_red_comp.rename(columns = {'Delayed':'NomDelays'}, inplace = True)
df_vuelos_red_comp

AcumDelay  NomDelays
UniqueCarrier                      
WN             1196574.0      34344
AA              936832.0      18066
UA              707809.0      13001
MQ              667065.0      13826
OO              624164.0      12625
XE              538530.0       9862
DL              487620.0      10612
CO              429869.0       8825
EV              406485.0       7966
YV              383108.0       6568
US              370092.0       8620
NW              350447.0       7521
B6              325477.0       5055
FL              324882.0       6747
OH              281150.0       5194
9E              249987.0       4979
AS              141833.0       3489
F9               83108.0       2684
HA               26631.0        739
AQ                1688.0         68

### Quins són els vols més llargs? I els més endarrerits? Busca les rutes més llargues i les que acumulen més retards.

In [86]:
df_vuelos_red[['Origin','Dest','Distance']].sort_values(by=['Distance'],ascending=False).drop_duplicates()

Origin Dest  Distance
176213     EWR  HNL      4962
1926816    HNL  EWR      4962
808619     HNL  ATL      4502
567559     ATL  HNL      4502
66596      ORD  HNL      4243
...        ...  ...       ...
1058279    GST  JNU        41
53134      SNA  LAX        36
53847      LAX  SNA        36
1239877    PSG  WRG        31
166946     WRG  PSG        31

[4815 rows x 3 columns]

In [87]:
df_vuelos_red[['Origin','Dest','ArrDelay']].groupby(['Origin', 'Dest']).sum().reset_index().sort_values(by=['ArrDelay'],ascending=False).query('ArrDelay > 0')

Origin Dest  ArrDelay
3468    ORD  LGA   27583.0
171     ATL  EWR   21874.0
2585    LGA  ORD   20559.0
2509    LAX  SFO   20249.0
3439    ORD  EWR   20064.0
...     ...  ...       ...
4120    SAV  BOS       1.0
4402    SLC  ELP       1.0
2028    IAD  TYS       1.0
2478    LAX  JAC       1.0
2621    LIH  PHX       1.0

[4770 rows x 3 columns]

### Aporta allò que consideris rellevant.

Buscamos si existe una correlación entre diferentes valores.

In [90]:
df_vuelos_red[['ArrDelay', 'Distance']].corr()

ArrDelay  Distance
ArrDelay  1.000000 -0.026421
Distance -0.026421  1.000000

In [94]:
df_vuelos_red[['Ave_Speed', 'DepDelay','ArrDelay' ]].corr()

Ave_Speed  DepDelay  ArrDelay
Ave_Speed   1.000000 -0.013598 -0.057536
DepDelay   -0.013598  1.000000  0.953034
ArrDelay   -0.057536  0.953034  1.000000

In [95]:
df_vuelos_red[['Distance', 'DepDelay','ArrDelay' ]].corr()

Distance  DepDelay  ArrDelay
Distance  1.000000 -0.005467 -0.026421
DepDelay -0.005467  1.000000  0.953034
ArrDelay -0.026421  0.953034  1.000000

Se puede observar que no existe una correlación entre los retrasos en llegada o salida, con la distancia del vuelo. Tampoco existe una relación entre la velocidad media del vuelo con el hecho de que existan retrasos en la salida, aunque la lógica parece indicar que para reducir este retraso sería lógico aumentar la velocidad del vuelo.
Si existe una correlación lógica entre los retrasos en salida y en llegada, aunque esto es un hecho evidente.

# Exercici 3

### Exporta el dataset net i amb les noves columnes a Excel.

In [96]:
df_vuelos_red.to_excel('Tasca_M3_T02.xlsx')